packages/create-cloudflare/src/templateMap.ts

```diff
@@ -1,53 +1,358 @@
+		return true;
+	}
+
+	// Otherwise, prompt the user for their TS preference
+	return processArgument<boolean>(ctx.args, "ts", {
+		type: "confirm",
+		question: "Do you want to use TypeScript?",
+		label: "typescript",
+		defaultValue: C3_DEFAULTS.ts,
+	});
+};
+
+export const processRemoteTemplate = async (args: Partial<C3Args>) => {
+	const templateUrl = await processArgument<string>(args, "template", {
+		type: "text",
+		question:
+			"What's the url of git repo containing the template you'd like to use?",
+		label: "repository",
+		validate: (val) => validateTemplateUrl(val || C3_DEFAULTS.template),
+		defaultValue: C3_DEFAULTS.template,
+	});
+
+	const path = await downloadRemoteTemplate(templateUrl);
+	const config = inferTemplateConfig(path);
+	validateTemplate(path, config);
+
+	return {
+		path,
+		...config,
+	};
+};
+
+const validateTemplate = (path: string, config: TemplateConfig) => {
+	if (typeof config.copyFiles?.path == "string") {
+		validateTemplateSrcDirectory(resolve(path, config.copyFiles.path), config);
+	} else {
+		for (const variant of Object.values(config.copyFiles as StaticFileMap)) {
+			validateTemplateSrcDirectory(resolve(path, variant.path), config);
+		}
+	}
+};
+
+const validateTemplateSrcDirectory = (path: string, config: TemplateConfig) => {
+	if (config.platform === "workers") {
+		const wranglerTomlPath = resolve(path, "wrangler.toml");
+		if (!existsSync(wranglerTomlPath)) {
+			crash(`create-cloudflare templates must contain a "wrangler.toml" file.`);
+		}
+	}
+
+	const pkgJsonPath = resolve(path, "package.json");
+	if (!existsSync(pkgJsonPath)) {
+		crash(`create-cloudflare templates must contain a "package.json" file.`);
+	}
+};
+
+/**
+ * Remote templates don't require a config file but may in the future. Until then, this
+ * function adapts a remote template to work with c3 by inferring a simple config from
+ * its file structure.
+ */
+const inferTemplateConfig = (path: string): TemplateConfig => {
+	return {
+		configVersion: 1,
+		id: "remote-template",
+		displayName: "A remote C3 template",
+		platform: "workers",
+		copyFiles: inferCopyFilesDefinition(path),
+	};
+};
+
+const inferCopyFilesDefinition = (path: string) => {
+	const copyFiles: StaticFileMap | VariantInfo = {};
+	if (existsSync(join(path, "js"))) {
+		copyFiles["js"] = { path: "./js" };
+	}
+	if (existsSync(join(path, "ts"))) {
+		copyFiles["ts"] = { path: "./ts" };
+	}
+	if (Object.keys(copyFiles).length !== 0) {
+		return copyFiles;
+	}
+
+	return {
+		path: ".",
+	};
+};
+
+/**
+ * Downloads an external template from a git repo using `degit`.
+ *
+ * @param src The url of the git repository to download the template from.
+ *            For convenience, `owner/repo` is also accepted.
+ * @returns A path to a temporary directory containing the downloaded template
+ */
+const downloadRemoteTemplate = async (src: string) => {
+	try {
+		const s = spinner();
+		s.start(`Cloning template from: ${blue(src)}`);
```

Comment:
This is very minor, but if I provide an invalid git repo, the loading dots remain in the output:

![Screenshot 2024-01-19 at 01 29 06](https://github.com/cloudflare/workers-sdk/assets/61631103/dfa7b53e-f826-44b4-9fec-1dd98e1ddde6)

![Screenshot 2024-01-19 at 01 28 57](https://github.com/cloudflare/workers-sdk/assets/61631103/31e65df6-ab52-403b-9ecc-4eba497d5573)

As I said it's pretty minor but I figured I might as well mention it 🙂